sd-2019-stack-overflow-frontend/src/model/questionModel.js

```diff
@@ -0,0 +1,136 @@
+            ],
+            newQuestion: createQuestion(4, "", "", "", "", [], 0),
+            newTag: "",
+            titleFilter: "",
+            tagFilter: "",
+            askQuestionModalActive: false
+        };
+    }
+
+    activateAskQuestionModal() {
+        this.state = {
+            ...this.state,
+            askQuestionModalActive: true
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    deactivateAskQuestionModal() {
+        this.state = {
+            ...this.state,
+            askQuestionModalActive: false
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    changeNewQuestionProperty(property, value) {
+        this.state = {
+            ...this.state,
+            newQuestion: {
+                ...this.state.newQuestion,
+                [property]: value
+            }
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    changeNewTag(newValue) {
+        this.state = {
+            ...this.state,
+            newTag: newValue
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    addNewTag(newTag) {
+        if (newTag === "") {
+            return;
+        }
+        this.state = {
+            ...this.state,
+            newQuestion: {
+                ...this.state.newQuestion,
+                tags: this.state.newQuestion.tags.concat([newTag])
+            },
+            newTag: ""
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    deleteNewTag(index) {
+        this.state = {
+            ...this.state,
+            newQuestion: {
+                ...this.state.newQuestion,
+                tags: this.state.newQuestion.tags.filter(t => t !== this.state.newQuestion.tags[index])
+            }
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    addQuestion(id, author, title, text, creationDateTime, tags, voteScore) {
+        this.state = {
+            ...this.state,
+            questions: [createQuestion(id, author, title, text, creationDateTime, tags, voteScore)].concat(this.state.questions),
+            newQuestion: createQuestion(id + 1, "", "", "", "", [], 0),
+            askQuestionModalActive: false
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    changeTitleFilter(newValue) {
+        this.state = {
+            ...this.state,
+            titleFilter: newValue
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    changeTagFilter(newValue) {
+        this.state = {
+            ...this.state,
+            tagFilter: newValue
+        };
+        this.emit("QuestionModelChange", this.state);
+    }
+
+    upvoteQuestion(questionId) {
+        var questions = [...this.state.questions];
+        var question = questions.find(q => q.id === questionId);
+        question.voteScore++;
```

Comment:
This changes the score inside the question (so it alters the state). You should clone the question as well :)

Reply:
Doesn't [...this.state.questions] clone the whole array and then I work on the cloned one?

Reply:
It clones the array, but not the objects inside the array (basically it creates a new array, containing the exact same references). See https://stackoverflow.com/questions/597588/how-do-you-clone-an-array-of-objects-in-javascript